src/lightgbm/src/main/scala/TrainUtils.scala

```diff
@@ -36,95 +36,172 @@ private object TrainUtils extends Serializable {
+              .map(j => lightgbmlib.doubleArray_getitem(evalResults, j)).toArray
+          } else if (iters - bestEval.get.bestIter(index) >= trainParams.earlyStoppingRound) {
+            isFinished = 1
+            log.info("Early stopping, best iteration is " + bestEval.get.bestIter(index))
+          }
         }
+        lightgbmlib.delete_doubleArray(evalResults)
+      }
+      iters = iters + 1
+    }
+  }
 
-        val isFinishedPtr = lightgbmlib.new_intp()
-        var isFinised = 0
-        var iters = 0
-        while (isFinised == 0 && iters < trainParams.numIterations) {
-          val result = lightgbmlib.LGBM_BoosterUpdateOneIter(boosterPtr.get, isFinishedPtr)
-          LightGBMUtils.validate(result, "Booster Update One Iter")
-          isFinised = lightgbmlib.intp_value(isFinishedPtr)
-          log.info("LightGBM running iteration: " + iters + " with result: " +
-            result + " and is finished: " + isFinised)
-          iters = iters + 1
+  def translate(labelColumn: String, featuresColumn: String, weightColumn: Option[String],
+                validationData: Option[Array[Row]], log: Logger,
+                trainParams: TrainParams, inputRows: Iterator[Row]): Iterator[LightGBMBooster] = {
+    if (!inputRows.hasNext)
+      List[LightGBMBooster]().toIterator
+    val rows = inputRows.toArray
+    var trainDatasetPtr: Option[SWIGTYPE_p_void] = None
+    var validDatasetPtr: Option[SWIGTYPE_p_void] = None
+    try {
+      val (trainRows, validRows) =
+        if (validationData.isDefined) {
+          (rows, Some(validationData.get))
+        } else {
+          (rows, None)
         }
-        val bufferLength = LightGBMConstants.defaultBufferLength
-        val bufferLengthPtr = lightgbmlib.new_longp()
-        lightgbmlib.longp_assign(bufferLengthPtr, bufferLength)
-        val bufferLengthPtrInt64 = lightgbmlib.long_to_int64_t_ptr(bufferLengthPtr)
-        val bufferOutLengthPtr = lightgbmlib.new_int64_tp()
-        val tempM =
-          lightgbmlib.LGBM_BoosterSaveModelToStringSWIG(
-            boosterPtr.get, 0, -1, bufferLengthPtrInt64,
-            bufferOutLengthPtr)
-        val bufferOutLength = lightgbmlib.longp_value(lightgbmlib.int64_t_to_long_ptr(bufferOutLengthPtr))
-        // TODO: Move the reallocation logic inside the SWIG wrapper
-        val model =
-          if (bufferOutLength > bufferLength) {
-            lightgbmlib.LGBM_BoosterSaveModelToStringSWIG(
-              boosterPtr.get, 0, -1, bufferOutLengthPtr, bufferOutLengthPtr)
-          } else tempM
-        log.info("Buffer output length for model: " + bufferOutLength)
+      trainDatasetPtr = generateDataset(trainRows, labelColumn, featuresColumn, weightColumn, None)
+      if (validRows.isDefined && validRows.get.nonEmpty) {
+        validDatasetPtr = generateDataset(validRows.get, labelColumn, featuresColumn, weightColumn, trainDatasetPtr)
+      }
+      var boosterPtr: Option[SWIGTYPE_p_void] = None
```

Comment:
please avoid vars if possible

Reply:
what would be the better way to write this bit?  I need to try/finally and cleanup in case something failed, but if createBooster itself throws I shouldn't free the booster.